README.md
# nhs-number

[![Clojars Project](https://img.shields.io/clojars/v/com.eldrix/nhs-number.svg)](https://clojars.org/com.eldrix/nhs-number)
[![cljdoc badge](https://cljdoc.org/badge/com.eldrix/nhs-number)](https://cljdoc.org/d/com.eldrix/nhs-number)
[![clj tests](https://github.com/wardle/nhs-number/actions/workflows/test-clj.yml/badge.svg)](https://github.com/wardle/nhs-number/actions/workflows/test-clj.yml)
[![cljs tests](https://github.com/wardle/nhs-number/actions/workflows/test-cljs.yml/badge.svg)](https://github.com/wardle/nhs-number/actions/workflows/test-cljs.yml)
[![codecov](https://codecov.io/gh/wardle/nhs-number/graph/badge.svg?token=K0NWSYYGZ8)](https://codecov.io/gh/wardle/nhs-number)

A Clojure/Script library providing validation, formatting and generation of UK NHS 
Numbers. As such, this library can be used for both front-end and back-end requirements.

The standards relating the NHS number are set out in [ISB 0149](https://digital.nhs.uk/data-and-information/information-standards/information-standards-and-data-collections-including-extractions/publications-and-notifications/standards-and-collections/isb-0149-nhs-number).
The NHS number is the unique patient identifier for a patient within the National Health 
Service in England and Wales.





# Development notes

## Automated GitHub actions

On every commit, an automated test suite is run using Clojure and ClojureScript.
On every tag, a release will be made automatically to Clojars.

## From the command-line

Run unit tests in Clojure:
```shell
clj -M:test:runner
```

Run unit tests in ClojureScript (this compiles the test source files into Javascript and runs using 'node.js'):
```shell
clj -M:test:node-cljs
```

Run unit tests in ClojureScript in a browser
```shell
clj -M:test:browser-cljs
```

A test server will be run which will return the results of the test suite.


Build a library jar file
```shell
clj -T:build jar
```

Deploy to Clojars repository: (requires valid `CLOJARS_USERNAME` and 
`CLOJARS_PASSWORD` environmental variables to be set)

```shell
clj -T:build deploy
```

# Other libraries

You can find similar libraries for other programming languages:

* Python : [https://github.com/uk-fci/nhs-number](https://github.com/uk-fci/nhs-number)


Mark Wardle


CHANGELOG.md
# Changes

This log documents significant changes for each release.

## [1.0.54] - 2023-10-16

* Add `valid*?` to normalise and validate in one convenient step
* Additional metadata including licence included in library jar build artefact
* Improvements to workflows in GitHub Actions
* Switch to Eclipse Public License (EPL) v2.0 

## [1.0.40] - 2023-08-18

* More complete test suite, run automatically using GitHub Actions, including coverage report via codecov.

## [1.0.8] - 2023-07-27

* Add support for ClojureScript. Can now be used in front-end applications 
as well as server-side.

## [1.0.4] - 2023-07-26

* Initial release with NHS number validation, formatting, normalisation and 
generation (migrated from [concierge](https://github.com/wardle/concierge))

test/com/eldrix/nhsnumber_test.clj
(ns com.eldrix.nhsnumber-test
  (:require [clojure.string :as str]
            [clojure.test :refer [deftest is run-tests]]
            [com.eldrix.nhsnumber :as nnn]))

(def valid-examples
  ["1111111111"
   "6328797966"
   "6148595893"
   "4865447040"
   "4823917286"])

(def invalid-examples
  [""
   " "
   "4865447041"
   "a4785"
   "1234567890"
   "111 111 1111"
   "          "])

(deftest test-valid
  (doseq [nnn valid-examples]
    (is (nnn/valid? nnn))))

(deftest test-invalid
  (doseq [nnn invalid-examples]
    (is (not (nnn/valid? nnn)))))

(deftest test-normalise
  (is (= (nnn/normalise "123 123 4567") "1231234567"))
  (is (nnn/valid? (nnn/normalise "111 111 1111"))))

(deftest test-format
  (is (= (nnn/format-nnn "1231234567") "123 123 4567")))

(deftest test-random-seq
  (let [xs (take 10000 (nnn/random-sequence 999))]
    (is (every? nnn/valid? xs))
    (is (every? true? (map #(str/starts-with? % "999") xs)))))

(deftest test-consecutive-seq
  (let [xs (take 10000 (nnn/ordered-sequence 999))]
    (is (every? nnn/valid? xs))
    (is (every? true? (map #(str/starts-with? % "999") xs)))))

(comment
  (run-tests))


build.clj
(ns build
  (:refer-clojure :exclude [compile])
  (:require [clojure.tools.build.api :as b]
            [deps-deploy.deps-deploy :as dd]))

(def lib 'com.eldrix/nhs-number)
(def version (format "1.0.%s" (b/git-count-revs nil)))
(def class-dir "target/classes")
(def jar-basis (b/create-basis {:project "deps.edn"}))
(def jar-file (format "target/%s-lib-%s.jar" (name lib) version))

(defn clean [_]
  (b/delete {:path "target"}))

(defn jar [_]
  (clean nil)
  (println "Building" jar-file)
  (b/write-pom {:class-dir class-dir
                :lib       lib
                :version   version
                :basis     jar-basis
                :src-dirs  ["src"]
                :scm       {:url                 "https://github.com/wardle/nhs-number"
                            :tag                 (str "v" version)
                            :connection          "scm:git:git://github.com/wardle/nhs-number.git"
                            :developerConnection "scm:git:ssh://git@github.com/wardle/nhs-number.git"}
                :pom-data  [[:description "A Clojure/Script library providing validation, formatting and generation of UK NHS Numbers"]
                            [:developers
                             [:developer
                              [:id "wardle"] [:name "Mark Wardle"] [:email "mark@wardle.org"] [:url "https://wardle.org"]]]
                            [:organization [:name "Eldrix Ltd"]]
                            [:licenses
                             [:license
                              [:name "Eclipse Public License v2.0"]
                              [:url "https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html"]
                              [:distribution "repo"]]]]})
  (b/copy-dir {:src-dirs ["src"], :target-dir class-dir})
  (b/copy-file {:src "LICENSE" :target "target/classes/META-INF/LICENSE"})
  (b/jar {:class-dir class-dir, :jar-file jar-file}))

(defn compile [_]
  (b/compile-clj {:basis        jar-basis
                  :src-dirs     ["src"]
                  :ns-compile   ['com.eldrix.nhsnumber]
                  :compile-opts {:elide-meta     [:doc :added]
                                 :direct-linking true}
                  :java-opts    ["-Dlogback.configurationFile=logback-build.xml"]
                  :class-dir    class-dir}))

(defn install
  "Installs pom and library jar in local maven repository"
  [_]
  (jar nil)
  (println "Installing" jar-file)
  (b/install {:basis     jar-basis
              :lib       lib
              :class-dir class-dir
              :version   version
              :jar-file  jar-file}))


(defn deploy
  "Deploy library to clojars.
  Environment variables CLOJARS_USERNAME and CLOJARS_PASSWORD must be set."
  [_]
  (println "Deploying" jar-file)
  (jar nil)
  (dd/deploy {:installer :remote
              :artifact  jar-file
              :pom-file  (b/pom-path {:lib       lib
                                      :class-dir class-dir})}))


src/com/eldrix/nhsnumber.clj
(ns com.eldrix.nhsnumber
  "Utility functions for validating, formatting, normalising and generating
   UK NHS numbers."
  (:require [clojure.string :as str]))

(set! *warn-on-reflection* true)

(defn ^:private check-digit
  "Calculate an NHS number check digit for the first nine characters of the
  string 's'. This is therefore usable for use on fully-formed 10-digit NHS
  numbers and for synthetically generated 9-digit NHS numbers. For the former,
  the calculated check digit should match the 10th digit, and the latter, the
  calculated check digit can be used to create a valid NHS number.
  Note: a check digit of 10 means that the NHS number is invalid."
  [s]
  (let [digits (map #(unchecked-subtract-int (int %) (int \0)) (str s)) ;; convert string into integers
        weights (range 10 1 -1)                             ;; the weights running from 10 down to 2
        total (reduce unchecked-add-int (map unchecked-multiply-int digits weights)) ;; multiply and total
        c1 (- 11 (mod total 11))]                           ;; calculate check digit
    (if (= 11 c1) 0 c1)))                                   ;; corrective fix when result is 11

(defn valid?
  "Validate an NHS number using the modulus 11 algorithm.
  An NHS number should be 10 numeric digits with the tenth digit a check digit.
  The validation occurs as follows:
  1. Multiply each of the first nine digits by a weighting factor (digit 1:10, 2:9, 3:8, 4:7, 5:6, 6:5, 7:4, 8:3, 9:2)
  2. Add the results of each multiplication together
  3. Divide total by 11, establish the remainder
  4. Subtract the remainder from 11 to give the check digit
  5. If result is 11, the check digit is 0
  6. If result is 10, NHS number is invalid
  7. Check remainder matches the check digit, if it does not NHS number is invalid"
  [^String nnn]
  (when (and (= 10 (count nnn)) (every? #(Character/isDigit ^char %) nnn))
    (let [cd (- (int (.charAt nnn 9)) (int \0))]            ;; the check digit
      (= cd (check-digit nnn)))))

(defn format-nnn
  "Formats an NHS number for display purposes into 3,3,4 format (as per standard
  ISB 0149 Amd 136/2010) with 'sep' used as separator (default, a blank space).
  e.g.,
  ```
  (format-nnn \"1112223304\")
  => \"111 222 3304\"
  ```"
  ([^String nnn]
   (format-nnn nnn " "))
  ([^String nnn sep]
   (if-not (= 10 (count nnn))
     nnn
     (str (subs nnn 0 3) sep (subs nnn 3 6) sep (subs nnn 6)))))

(defn normalise
  "Normalise an NHS number, removing non-digit characters such as whitespace and
  punctuation. Returns 'nil' if the resulting string is not 10 digits long.
  e.g.,
  ```
  (normalise \"111 222 3304\")
  => \"1112223304\"
  ```"
  [s]
  (let [s' (str/replace s #"\D" "")]
    (when (= 10 (count s')) s')))

(def ^:private xf-generate
  "A transducer that transforms input strings by padding to 9 digits, adding a
  check digit and removing invalid generated NHS numbers."
  (comp
    (map #(let [s (format "%09d" %)] (str s (check-digit s))))                    ;; this *can* generate invalid NHS numbers (e.g. if check digit is 10)
    (filter valid?)))

(defn ^:private nnn-range
  "Given a 'prefix' generate a range as a vector of start and end. Prefix can
  be a string, or number.
  e.g.,
  ```
  (nnn-range 56)
  => [560000000 569999999]
  ```"
  [prefix]
  (let [n (count (str prefix))]
    (when (> n 9)
      (throw (ex-info (str "Invalid prefix: " prefix) {})))
    (vector
      (or (parse-long (apply str prefix (repeat (- 9 n) \0))) 0)
      (or (parse-long (apply str prefix (repeat (- 9 n) \9))) 0))))

(defn ^:private rand-int-range
  "Returns a random integer between start and end (inclusive)."
  [start end]
  (+ start (rand-int (- (inc end) start))))

(defn random-sequence
  "Returns a lazy sequence of randomly generated valid NHS numbers. This may
  generate duplicate NHS numbers within the sequence. If only distinct NHS
  numbers are required, use:
  ```
  (take 50 (distinct (random-sequence 999)))
  ```
  This example generates 50 unique NHS numbers with the prefix '999'."
  ([]
   (random-sequence nil))
  ([prefix]
   (let [[start end] (nnn-range prefix)]
     (sequence xf-generate (repeatedly #(rand-int-range start end))))))

(defn random
  "Generate a random valid NHS number with the given prefix. There are no
  guarantees that duplicate NHS numbers will not be generated on successive
  calls. As such, if a sequence of unique NHS numbers are required, use:
  ```
  (take 50 (distinct (random-sequence \"999\")))
  ```
  Alternatively, store previously used generated NHS numbers and retry if a
  duplicate is generated."
  ([]
   (random nil))
  ([prefix]
   (first (random-sequence prefix))))

(defn ordered-sequence
  "Return a lazy sequence of ordered, valid NHS numbers. Given the results will
  be ordered, they are guaranteed to be unique within the same sequence."
  ([]
   (ordered-sequence nil))
  ([prefix]
   (let [[start end] (nnn-range prefix)]
     (sequence xf-generate (range start end)))))

(comment
  (valid? "1111111111")
  (check-digit "1111111111")
  (take 50 (ordered-sequence "999"))
  (take 50 (distinct (random-sequence "999")))
  (require '[criterium.core :as crit])
  (format-nnn "1234567890")
  (normalise "123 456 7890")
  (crit/bench (format-nnn "11111111111")))




